boards: arm: Add QEMU Cortex-M4 emulation platform

stephanosio opened the pull request:
This commit adds the QEMU Cortex-M4 emulation platform (qemu_cortex_m4)
based on the new `netduinoplus2` machine added in the QEMU 5.0.

The QEMU `netduinoplus2` machine emulates the Netduino Plus 2 hardware
that includes the STM32F405 SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

Closes #22870

NOTE:
1. This PR must be merged after the QEMU 5.0 is released, integrated into the Zephyr SDK (zephyrproject-rtos/sdk-ng#192), and the CI is updated to use the new SDK.
2. In order to test this PR, follow the instructions below:
    1. Check out the following branch (which has all the necessary patches pre-applied):
        https://github.com/stephanosio/qemu/commits/zephyr_stm32f405_emu
    2. Build QEMU `arm-softmmu` target (`configure --target-list=arm-softmmu` `make`)
    3. Set`QEMU_BIN_PATH` to the directory that contains the `qemu-system-arm` built above.

TODO:
- [x] Confirm that this PR passes all tests with https://github.com/zephyrproject-rtos/sdk-ng/pull/201
- [x] Enable icount and confirm that this PR passes all tests

stephanosio committed:
boards: arm: Add QEMU Cortex-M4 emulation platform

This commit adds the QEMU Cortex-M4 emulation platform (qemu_cortex_m4)
based on the new `netduinoplus2` machine added in the QEMU 5.0.

The QEMU `netduinoplus2` machine emulates the Netduino Plus 2 hardware
that includes the STM32F405 SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
dts: bindings: vendor-prefixes: Add Wilderness Labs

Add Wilderness Labs to the vendor prefixes list

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg commented on boards/arm/qemu_cortex_m4/doc/index.rst:
```diff
@@ -0,0 +1,102 @@
+
+* Nested Vectored Interrupt Controller
+* System Tick System Clock
+* Stellaris UART
```
Should we add the floating-point co-processor?

stephanosio replied:
Done

ioannisg reviewed (commented):
The PR looks good to me, I 've got a minor comment only in the .rst

stephanosio commented:
Rebased

stephanosio committed:
boards: arm: qemu_cortex_m4: Disable hwinfo driver

This commit disables the STM32 hwinfo driver for the QEMU Cortex-M4
testing platform, as the UID register (at 0x1FFF7A10) is not
implemented on the QEMU.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Verified that all tests pass for `qemu_cortex_m4` using the [patched QEMU 5.0.0 release](https://github.com/zephyrproject-rtos/sdk-ng/pull/201):
https://gist.github.com/stephanosio/70d203d98a4086068b13887e6fe89249

ioannisg reviewed (commented):
Have you thought naming this qemu_cortex_m4f? That is, indicate that this includes the floating-point coverage, which will be unique among Qemu ARM platforms.

stephanosio commented:
Rebased

ioannisg commented on boards/arm/qemu_cortex_m4f/qemu_cortex_m4f_defconfig:
```diff
@@ -0,0 +1,27 @@
+CONFIG_CLOCK_CONTROL=y
+
+# Enable FPU
+CONFIG_FPU=y
```
Probably not needed, since the respective FP tests enable this, but i guess there's no harm.

ioannisg commented on boards/arm/qemu_cortex_m4f/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+config HWINFO_STM32
+	default n
+
+endif # BOARD_QEMU_CORTEX_M4
```
nit: BOARD_QEMU_CORTEX_M4F

ioannisg reviewed (approved):
Great work @stephanosio this looks fine to me. Left two minor comments that you could address.
Happy to be able to, finally, see Floating Point coverage in QEMU.

bwhitman commented:
@stephanosio , hi! I'm very interested in getting this running to test. I applied your patches, and am having a `qemu: fatal: Lockup: can't escalate 3 to HardFault (current priority -1)` error while running `hello_world`. Any advice? 

Here we can see cortex-m3 works fine but the cortex-m4f gets the fault. 
```
bwhitman@ubuntu:~$ cd outside/zephyrproject/zephyr/samples/hello_world/
bwhitman@ubuntu:~/outside/zephyrproject/zephyr/samples/hello_world$ rm -rf build; mkdir build; cd build; cmake -GNinja -DBOARD=qemu_cortex_m3 ..; ninja run

[...]

*** Booting Zephyr OS build zephyr-v2.3.0-2014-g127e82e3b80c  ***
Hello World! qemu_cortex_m3
QEMU: Terminated
bwhitman@ubuntu:~/outside/zephyrproject/zephyr/samples/hello_world/build$ cd ..
bwhitman@ubuntu:~/outside/zephyrproject/zephyr/samples/hello_world$ export QEMU_BIN_PATH=/home/bwhitman/qemu-5.0.0/arm-softmmu/
bwhitman@ubuntu:~/outside/zephyrproject/zephyr/samples/hello_world$ rm -rf build; mkdir build; cd build; cmake -GNinja -DBOARD=qemu_cortex_m4f ..; ninja run
[...]
[QEMU] CPU: cortex-m4
qemu: fatal: Lockup: can't escalate 3 to HardFault (current priority -1)

R00=ffffffe0 R01=00000000 R02=fffffff9 R03=00000000
[...]
Aborted (core dumped)
```

bwhitman commented:
Never mind, I missed that I needed your [particular commit of qemu](https://github.com/stephanosio/qemu/commits/zephyr_stm32f405_emu). It works now! Thank you! 

``` CPU: cortex-m4
*** Booting Zephyr OS build zephyr-v2.3.0-2014-g127e82e3b80c  ***
Hello World! qemu_cortex_m4f```

stephanosio commented:
@bwhitman Correct, it requires some patches that are not included in the SDK.

Speaking of which, I wonder if we will ever get to merging them ...
